File: 202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = getNext(n);

        while(fast != 1 && slow != fast) {
            slow = getNext(slow);
            fast = getNext(getNext(fast));
        }
        
        return fast == 1;        
    }
    
    /*public boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();

        while(n != 1 && !seen.contains(n)) {
            seen.add(n);
            n = getNext(n);
        }
        
        return n == 1;        
    }*/
    
    private int getNext(int n) {
        int sum = 0;
        while(n > 0) {
            int temp = n % 10;
            sum += temp * temp;
            n /= 10;
        }

        return sum;
    }
}

File: top150/202_happy_number.py
# Method 1: slow fast pointer
class Solution:
    def isHappy(self, n: int) -> bool:
        slow = self.getNext(n)
        fast = self.getNext(slow)
        while fast != slow:
            slow = self.getNext(slow)
            fast = self.getNext(self.getNext(fast))
        
        return fast == 1
    
    def getNext(self, n):
        total = 0
        while n != 0:
            lastDigit = n % 10
            total += lastDigit * lastDigit
            n //= 10
        return total


# Method 2: Use hash map
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()
        while n not in seen:
            seen.add(n)
            n = self.getNext(n)
        
        return n == 1
    
    def getNext(self, n):
        total = 0
        while n != 0:
            lastDigit = n % 10
            total += lastDigit * lastDigit
            n //= 10
        return total
